p2p/src/network.rs

```diff
@@ -2,516 +2,451 @@
-        oneshot::{Receiver, Sender},
-    },
+    sync::{mpsc, oneshot, watch},
 };
 
 use crate::{
-    peer::{Peer, PeerId},
+    boilerplate::*,
+    peer::{
+        handles::{connected_from, connecting, PeerHandle},
+        message::*,
+        Connection, ConnectionId,
+    },
     Error,
 };
 
-/// Reference as a means of communication with a [`Peer`]
-#[derive(Clone, Debug)]
-pub struct RefPeer<T, K, E>
-where
-    T: Debug + Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
-    K: KeyExchangeScheme + Send + 'static,
-    E: Encryptor + Send + 'static,
-{
-    addr: Addr<Peer<T, K, E>>,
-    conn_id: ConnectionId,
-    p2p_addr: String,
+/// Handle to interact with [`NetworkBase`] actor.
+#[derive(derive_more::DebugCustom)]
+#[debug(fmt = "core::any::type_name::<Self>()")]
+pub struct NetworkBaseHandle<T: Pload, K: Kex, E: Enc> {
+    /// Sender to subscribe for message `T`
+    subscribe_t_sender: mpsc::Sender<mpsc::Sender<T>>,
+    /// Receiver of `OnlinePeer` message
+    online_peers_receiver: watch::Receiver<OnlinePeers>,
+    /// [`ConnectPeer`] message sender
+    connect_peer_sender: mpsc::Sender<ConnectPeer>,
+    /// [`DisconnectPeer`] message receiver
+    disconnect_peer_sender: mpsc::Sender<DisconnectPeer>,
+    /// Sender of [`Post`] message
+    post_sender: mpsc::Sender<Post<T>>,
+    /// Key exchange used by network
+    _key_exchange: core::marker::PhantomData<K>,
+    /// Encryptor used by the network
+    _encryptor: core::marker::PhantomData<E>,
 }
 
-/// Base network layer structure, holding connections, called
-/// [`Peer`]s.
-pub struct NetworkBase<T, K, E>
-where
-    T: Debug + Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
-    K: KeyExchangeScheme + Send + 'static,
-    E: Encryptor + Send + 'static,
-{
-    /// Listening address for incoming connections. Must parse into [`std::net::SocketAddr`]
-    listen_addr: String,
-    /// [`Peer`]s performing a `handshake`
-    pub new_peers: HashMap<ConnectionId, Addr<Peer<T, K, E>>>,
-    /// Current [`Peer`]s in [`Peer::Ready`] state.
-    pub peers: HashMap<PublicKey, RefPeer<T, K, E>>,
-    /// [`HashSet`] of [`String`] which should represent the [`std::net::IpAddr`] of the untrusted remote [`Peer`]:
-    /// inserted by [`DisconnectPeer`] and removed by [`ConnectPeer`] from Sumeragi.
-    /// In case the [`String`] represents an unresolved hostname, the first reconnection is not refused
-    untrusted_peers: HashSet<String>,
-    /// [`TcpListener`] that is accepting [`Peer`]s' connections
-    pub listener: Option<TcpListener>,
-    /// Our app-level public key
-    public_key: PublicKey,
-    /// [`iroha_actor::broker::Broker`] for internal communication
-    pub broker: Broker,
-    /// Flag that stops listening stream
-    finish_sender: Option<Sender<()>>,
-    /// Buffer capacity of actor's MPSC channel
-    actor_channel_capacity: u32,
+impl<T: Pload, K: Kex, E: Enc> Clone for NetworkBaseHandle<T, K, E> {
```

Comment:
Unfortunately, this won't work, because `derive(Clone)` infer bounds incorrectly (PhantomData already implement `Copy` itself). 
You can take a look at this [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c9509b655061065e32f9bf498e719045).